### PR TITLE
Cache Azure credential obtained from environment

### PR DESCRIPTION
**Pull request by jyemin**
JAVA-4706

### CODE REVIEW

File: driver-core/src/main/com/mongodb/internal/ExpirableValue.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import static com.mongodb.assertions.Assertions.assertNotNull;
+import static com.mongodb.internal.VisibleForTesting.AccessModifier.PRIVATE;
+
+/**
+ * A value associated with a lifetime.
+ *
+ * <p>Instances are shallowly immutable.</p>
+ */
+@ThreadSafe
+public final class ExpirableValue<T> {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/ExpirableValue.java:33]: Most of the logic is in this class, which is unit tests.  Thanks @stIncMale for the suggestion

File: driver-core/src/main/com/mongodb/internal/authentication/AzureCredentialHelper.java [hunk 2]
```diff
@@ -31,25 +35,44 @@
...
+    private static final String ACCESS_TOKEN_FIELD = "access_token";
+    private static final String EXPIRES_IN_FIELD = "expires_in";
+
+    private static ExpirableValue<String> cachedAccessToken = ExpirableValue.expired();
+
+    public static synchronized BsonDocument obtainFromEnvironment() {
+        String accessToken;
+        Optional<String> cachedValue = cachedAccessToken.getValue();
+        if (cachedValue.isPresent()) {
+            accessToken = cachedValue.get();
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/authentication/AzureCredentialHelper.java:47]: This is the only untested line

File: driver-core/src/main/com/mongodb/internal/authentication/AzureCredentialHelper.java [hunk 3]
```diff
@@ -31,25 +35,44 @@
...
+
+            if (!responseDocument.isString(ACCESS_TOKEN_FIELD)) {
+                throw new MongoClientException(String.format(
+                        "The %s field from Azure IMDS metadata response is missing or is not a string", ACCESS_TOKEN_FIELD));
+            }
+            if (!responseDocument.isString(EXPIRES_IN_FIELD)) {
+                throw new MongoClientException(String.format(
+                        "The %s field from Azure IMDS metadata response is missing or is not a string", EXPIRES_IN_FIELD));
             }
-        } catch (JsonParseException e) {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/authentication/AzureCredentialHelper.java:50]: It reads cleaner with the try catch having a narrower scope

File: driver-core/src/main/com/mongodb/internal/authentication/AzureCredentialHelper.java [hunk 4]
```diff
@@ -31,25 +35,44 @@
...
-            throw new MongoClientException("Exception parsing JSON from Azure IMDS metadata response.", e);
-        }
+            accessToken = responseDocument.getString(ACCESS_TOKEN_FIELD).getValue();
+            int expiresInSeconds = Integer.parseInt(responseDocument.getString(EXPIRES_IN_FIELD).getValue());
+            cachedAccessToken = ExpirableValue.unexpired(accessToken, Duration.ofSeconds(expiresInSeconds).minus(Duration.ofMinutes(1)));
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/authentication/AzureCredentialHelper.java:73]: Create the new cached access token with the expiration as the duration from the HTTP response minus one minute

**Review by stIncMale — APPROVED** [review]

File: driver-core/src/main/com/mongodb/internal/authentication/AzureCredentialHelper.java [hunk 5]
```diff
@@ -31,25 +35,44 @@
...
+            if (!responseDocument.isString(EXPIRES_IN_FIELD)) {
+                throw new MongoClientException(String.format(
+                        "The %s field from Azure IMDS metadata response is missing or is not a string", EXPIRES_IN_FIELD));
             }
-        } catch (JsonParseException e) {
-            throw new MongoClientException("Exception parsing JSON from Azure IMDS metadata response.", e);
-        }
+            accessToken = responseDocument.getString(ACCESS_TOKEN_FIELD).getValue();
+            int expiresInSeconds = Integer.parseInt(responseDocument.getString(EXPIRES_IN_FIELD).getValue());
+            cachedAccessToken = ExpirableValue.expirable(accessToken, Duration.ofSeconds(expiresInSeconds).minus(Duration.ofMinutes(1)));
```
Comment (DmitryLukyanov) [inline driver-core/src/main/com/mongodb/internal/authentication/AzureCredentialHelper.java:73]: the spec wording about this logic says: https://github.com/mongodb/specifications/commit/d6b8cce6abb3b8e1a0b8f1dc7ee737e18322cfce#diff-de01d398114f63480af23798e09ffca599aa758efd04f15a9beec89b4566041bR591 a bit different calculating logic. When we calculate this value:
>ExpirableValue.expirable(accessToken, Duration.ofSeconds(expiresInSeconds).minus(Duration.ofMinutes(1)));

it's expected to add `expire_in` not to the current time, but to the time when acquiring credentials had been starting (`t_0` is saved before http call). It's supposed protecting us against very long aws request processing. I think it's in 99% cases is overkill, but It's current spec wording.

**Review by DmitryLukyanov — COMMENTED** [review]
See minor comment about how calculate expired date. Then it will be LGTM

**Review by DmitryLukyanov — APPROVED** [review]
LGTM

**Review by stIncMale — APPROVED** [review]

File: driver-core/src/main/com/mongodb/internal/authentication/AzureCredentialHelper.java (hunk 5 above)
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/internal/authentication/AzureCredentialHelper.java]: Given that both Jeff and Dima are unhappy with "Done", I feel like I now need to continue the trend: ☹

Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/authentication/AzureCredentialHelper.java]: Ack ;)

**Pull request merged**